### PR TITLE
update: Parallel DB locks watcher

### PR DESCRIPTION
**Pull request by Yajo**
With this patch, if you update a database while another Odoo instance is running (such as i.e. a production instance), the production instance will not be locked just because there's a DB lock.

DB locks can happen i.e. when 2 or more transactions are happening in parallel and one of them wants to modify data in a field while another one is modifying the field itself. For example, imagine that a user is modifying a `res.partner`'s name, while another update process is adding a `UNIQUE` constraint in the `name` field of the `res_partner` table. This would produce a deadlock where each transaction is waiting for the other one to finish, and thus both the production instance and the update instance would be locked infinitely until one of them is aborted.

You cannot detect such problem with common tools such as timeouts, because there still exists the possibility of a query being slow without actually being locked, like when you update an addon that has a pre-update migration script that performs lots of work, or when your queries or server are not optimized and perform slowly.

So, the only way to detect deadlocks is by issuing a separate DB cursor that is not protected by a transaction and that watches other cursors' transactions and their locks.

With this change, this is what happens now behind the scenes:

- The DB update process is spawned in background.
- The foreground process uses a separate watcher cursor and watches for locks.
- If a lock is detected, the update process is aborted, giving priority to the other cursors. This is by design because your production users have priority always, and all that would happen is that the update transaction would be rolled back, so you can just try updating again later.
- A couple of CLI parameters allow you to modify the watcher behavior.

---

BTW, the changes in pre-commit are because I have Fedora, where the py3 version is 3.7. I guess this change will help other contributors too, to avoid pre-commit failures...

@Tecnativa

### CODE REVIEW

**Comment by Yajo** [comment]
Please set as WIP for now.

**Comment by Yajo** [comment]
This shouldn't be WIP now, it's ready to review.

I published https://youtu.be/20bUjtO_Mjk for you to be able to see what is this doing in practice. Basically, it wants to allow safely updating in parallel with a production running database, while not disturbing for other use cases.

I'd appreciate some help on putting :heavy_check_mark: travis, because the errors seem not related or at least I don't understand the logs.

Thanks.

**Comment by blaggacao** [comment]
First here is what I understand of the spec:

- Be able to update a production db while it is running in production.

I'm not the code owner, but I'm a interested third party. Therefore, going over that spec, first:

- Is running updates on live production database instances really a strategy you want to employ (as opposed to reusing a maintenance windows for as short and ad hoc as it might be)

Please correct me if I misunderstood anything.

**Comment by Yajo** [comment]
Yes, that's the purpose. I don't want to ask my customers if I can update, I want to update at any time without them noticing at all. HA, basically. Of course, this feature must be combined with some kind of rolling update system to be helpful, but that's outside this PR's scope.

**Comment by Yajo** [comment]
I can't understand Travis... Where's the failure? Thanks!

**Comment by sbidoul** [comment]
I don't understand it either... no time to dig into this now unfortunately.

**Comment by lmignon** [comment]
Strange, but without more details into the log, it's not possible to understand the problem. Since the error occurs into the code I've written, I'll try to find some time to dig into.... (but not today...)

**Comment by Yajo** [comment]
I guess we should increase logging in Travis. I can try if you tell me how.

**Comment by blaggacao** [comment]
@Yajo Got it. It looks like a two-sided sword to me. And I'm not sure if that would convince me of being the right approach. I repeat, I'm not the code owner, so I'm just a vocal spectator. You can just ignore me if you want. :wink: 

But still it's valuable to understand your thought process. For example, I cannot really figure out what would happen if an update transaction would update the schema while another user transaction still relies on a previous schema state. Which of both transactions would corrupt and fail? Is that possible at all?

Please share your insights. :wink: 

As for now and current state of (my) knowledge, it seems to me that updates without schema changes could be executed safely. The problem I see is that this feature of an update is opaque during the odoo update process: We cannot scope such "hot" updates to non-schema-changes.

I might just as likely be completely wrong in my analysis. Please let me know! :smile:

**Comment by Yajo** [comment]
> what would happen if an update transaction would update the schema while another user transaction still relies on a previous schema state

Well, that's precisely the case we're handling here.

What would happen is that both cursors would become locked, in an endless loop where each one is waiting for the other one to finish. You can see a little example in the video above.

We have been updating in parallel for years now, and it works pretty well unless this precise situation happens, because most updates do not alter the schema, or at least not while a user is specifically using a previous schema in the precise field or table that is being altered.

The solution we have developed for this problem is to just keep our eyes peeled over the logs, and if they become stuck, abort and restart the update process. It is far from ideal or maintainable! This patch is replacing our eyes by python code that spawns a watcher for locks and aborts the update if they happen.

So, while you'd still be able to _stop :arrow_right: update :arrow_right: start_ as before, now you are able to _update (:repeat: repeat if failed) :arrow_right: restart_.

The final update time can be higher, but the downtime will be almost nothing.

---

As a side note, for you to understand better, production Odoo process would be running in a different codebase than the update process. You can use separate installations or containers for such purpose. For instance, using containers:

1. Production container _PCA_ based on image _IA_ is running Odoo in production.
1. Here you'd trigger a backup, just in case.
1. A new update container _UCB_ based on image _IB_ is spawned, and it runs `click-odoo-update` against the same database instance as _PCA_.
1. If _UCB_ exits with failure, update has failed. Repeat step 3 as many times as needed.
1. If _UCB_ exits normally, update has finished properly. Replace _PCA_ with a new container _PCB_, based on image _IB_.

The only downtime users would experience is in the last step, but even that can be removed by using more replicas for _PCA_ and replacing them gradually. Of course this is not a one-size-fits-all solution, but at least it opens the door for real HA in Odoo, which has been closed for the last centuries.

**Comment by blaggacao** [comment]
@Yajo Understanding what you said, then, I guess this line is utterly confusing: https://github.com/acsone/click-odoo-contrib/pull/38/files#diff-b241a31b4530ee322670fff4974f8be1R33

> Watch DB while another process is updating Odoo"

Maybe you'd say: "Watch DB for existing locks while this process tries to update Odoo"

Would that be more assertive?

As for my motivation:

In Odoo Operator I'm implementing a slightly different workflow:

1. Trigger "Stage Migration"
2. Backup A
3. Copy A -> B
4. Migrate B.
5. Expose B to the client for validation
6. Trigger "Migration"
7. Set A in maintenance (splash screen)
8. Backup A
9. Copy A -> B
10. Migrate B
11. Permutade A / B wirings at all necesary levles 

For sure almost every task is executed as a separate job (or container, if you want).

If a version bump object is marked as `bugfix=true`, then it would auto-apply, which means

1. Backup A
2. Migrate A
3. Done

I think your idea is interesting for very short "bugfixing" workflows to cover those corner cases.

I'm still not fully understanding how / if that benefits since the migration is not split up into several transaction, so the transaction span is really the TTL of the migration and for the migrations to succeed it would need to block other users (which is down time).

Or maybe I did not understand this part: 

- Is your strategy to retry migrations until there is a sufficiently large time span for the migration to succeed and always give preference to user connections?

**Comment by blaggacao** [comment]
@Yajo If I was more processional than I actually am, I'd probably just ask would you mind sharing a spec? :wink:

**Comment by Yajo** [comment]
About your 1st comment, I see you answered everything yourself. About the 2nd... :grin: If I were more professional I'd understand what a _spec_ is!

In any case, the point of this PR is just opening a new door for us. You still are able to work as before, changing nothing (because the watcher would never hit a lock anyways, if you don't update in parallel), so no matter if you use Operator, Doodba, Ansible or Windows 98, this is just a new door you can choose to use if you want and know the implications.

I think that we shouldn't clutter this PR anymore with specific implementations. :thinking:  That discussion fits better in downstream infrastructure projects, don't you think?

---

I see that I should update the README too. I'd like someone  from @acsone to help me before with that nasty :x: though.

**Comment by lmignon** [comment]
@Yajo For sure, I'll check what's wrong with travis in the next days...  :smirk: I keep you informed once I've fixed the problem.

File: click_odoo_contrib/update.py [hunk 1]
```diff
@@ -88,13 +198,46 @@ def OdooEnvironmentWithUpdate(database, ctx, **kwargs):
                 "Updating addons for their hash changed: %s.",
                 ",".join(to_update.keys()),
             )
-    if ctx.params["i18n_overwrite"]:
-        odoo.tools.config["overwrite_existing_translations"] = True
+    odoo.tools.config["overwrite_existing_translations"] = ctx.params["i18n_overwrite"]
```
Comment (sbidoul) [inline click_odoo_contrib/update.py]: @Yajo not sure about this one. Can you make another PR for this line?

File: click_odoo_contrib/update.py [hunk 2]
```diff
@@ -88,13 +198,46 @@ def OdooEnvironmentWithUpdate(database, ctx, **kwargs):
...
     if odoo.release.version_info[0] < 10:
         Registry = odoo.modules.registry.RegistryManager
     else:
         Registry = odoo.modules.registry.Registry
-    Registry.new(database, update_module=True)
+    try:
+        # Update Odoo spawning a different process
+        update_process = multiprocessing.Process(
+            target=Registry.new, kwargs={"db_name": database, "update_module": True}
+        )
```
Comment (sbidoul) [inline click_odoo_contrib/update.py]: To be less intrusive, would it be possible to run the watcher in a subprocess (or thread) only if activated by an option, and let the update run in the main process in all cases?

Comment (Yajo) [reply click_odoo_contrib/update.py]: I thought about that option, but let me share my thoughts on the matter:

1. [Only the process that spawned another process is able to join or terminate it](https://docs.python.org/2.7/library/multiprocessing.html#multiprocessing.Process.terminate). If, at some point, Odoo is able to properly handle signals while locked and reap its cursors, it'd be even better to terminate this process top-down instead of bottom-up as it is right now (due to this restriction). In such case, we'd have to terminate it from the main process anyways.
1. Handling 2 subprocesses and making one watch over the other is way harder than with only 1 subprocess. At the end of the day, we'd have to:

   - spawn update
   - spawn watcher
   - do a `while True` loop until both finish

   It's more complex than the current, but at the end of the day would do the same. Current:

   - spawn update
   - watch in a `while True` loop
1. I could add an option that avoids the watcher, and in such case we could avoid spawning processes and behave as before, but after all the watcher is supposed to detect 0 locks if we are not updating in parallel, so at the end of the day nobody would notice it if not using parallel updates, and we'd have to maintain 2 ways of using the script instead of just 1 that, although is more complex, covers both scenarios.

Please let me know what you think about it. :blush:

File: click_odoo_contrib/update.py (hunk 1 above)
Comment (Yajo) [reply click_odoo_contrib/update.py]: Yes, this was more a "BTW" than anything else.

File: click_odoo_contrib/update.py (hunk 2 above)
Comment (sbidoul) [reply click_odoo_contrib/update.py]: I was thinking about something like this:
- launch watcher in a daemon thread
- main process does the update
- when watcher detects deadlocks it terminates the postgres backend (which in turn should raise an exception that terminates the main update process)
- the daemon watcher thread is killed automatically when the main process exits

Comment (Yajo) [reply click_odoo_contrib/update.py]: OK, that makes more sense, but which would be the benefits?

Comment (sbidoul) [reply click_odoo_contrib/update.py]: @Yajo the change is less intrusive (the normal flow stays unchanged) and you can start the watcher thread only when activated with an option. So less risk and the code would be easier to understand (by me at least) and maintain.

Comment (Yajo) [reply click_odoo_contrib/update.py]: I don't think we should add an option to trigger the watcher... :thinking: 

If you don't update in parallel (AFAIK you don't do this actually), the watcher will basically do nothing.

If you do update in parallel, then you always need the watcher.

So, what do we gain on adding an option? Isn't it better to have a watcher that just works always?

Comment (sbidoul) [reply click_odoo_contrib/update.py]: Well, the watcher will kill the update if a query lasts for more than 30 seconds by default. It's an advanced feature that can have unexpected effects, so its better to let advanced users activate it when they understand they need it.

Comment (Yajo) [reply click_odoo_contrib/update.py]: > the watcher will kill the update if a query lasts for more than 30 seconds

Actually not. It will kill the update _if there's a query locking another query_ for more than 30 seconds.

A single process would never have 2 queries in parallel, and even less 2 blocking ones, so it should never affect.

This situation can _only_ happen when there are more than 1 clients connected.

**Comment by Yajo** [comment]
Sorry for the noise, please set as wip again. I'm fixing some rough edges.

File: click_odoo_contrib/update.py [hunk 3]
```diff
@@ -97,6 +184,26 @@ def OdooEnvironmentWithUpdate(database, ctx, **kwargs):
...
+    # Watch for database locks while Odoo updates
+    watcher_process = multiprocessing.Process(
+        target=_db_lock_watcher,
+        args=(database, ctx.params["watcher_beat"], ctx.params["watcher_max_seconds"]),
+    )
+    try:
+        if ctx.params["watcher_max_seconds"] > 0:
+            watcher_process.start()
+        # Update Odoo datatabase
+        _update_db(database, ctx.params["update_all"], ctx.params["i18n_overwrite"])
```
Comment (Yajo) [inline click_odoo_contrib/update.py]: The update process is moved to its own method, so code in this method looks more clean because all lit does it to handle both the update and the watcher methods..

File: click_odoo_contrib/update.py [hunk 4]
```diff
@@ -97,6 +184,26 @@ def OdooEnvironmentWithUpdate(database, ctx, **kwargs):
...
+    finally:
+        if watcher_process.is_alive():
+            _logger.info("Terminating DB lock watcher")
+            watcher_process.terminate()
+            watcher_process.join()
```
Comment (Yajo) [inline click_odoo_contrib/update.py]: The watcher termination is simpler because it should always work.

File: click_odoo_contrib/update.py [hunk 5]
```diff
@@ -113,10 +220,26 @@ def OdooEnvironmentWithUpdate(database, ctx, **kwargs):
...
+    "--watcher-beat",
+    default=10.0,
+    type=float,
+    help="Time to wait between DB lock watcher loops",
+)
+@click.option(
+    "--watcher-max-seconds",
+    default=30.0,
+    type=float,
+    help="Max DB lock seconds allowed before aborting the update process. 0=disable",
```
Comment (Yajo) [inline click_odoo_contrib/update.py]: You can skip spawning the watcher by setting `--watcher-max-seconds=0`.

I think that to leave the watcher on by default is safer since it covers more use cases, but I left the option to skip it so you can feel safer about it. :blush:

File: click_odoo_contrib/update.py (hunk 4 above)
Comment (sbidoul) [reply click_odoo_contrib/update.py]: If you start the watcher as thread, there is no need for code to terminate it.

File: click_odoo_contrib/update.py (hunk 5 above)
Comment (sbidoul) [reply click_odoo_contrib/update.py]: I insist to have the watcher disabled by default.

Comment (Yajo) [reply click_odoo_contrib/update.py]: Done.

File: click_odoo_contrib/update.py (hunk 4 above)
Comment (Yajo) [reply click_odoo_contrib/update.py]: I prefer to use processes, to make sure they are not blocked by the GIL. Terminating it is not so hard.

Comment (sbidoul) [reply click_odoo_contrib/update.py]: Or you can do `watcher_process.daemon = True` so termination will be automatic.

A thread is still better in this case: no risk of leaving processes hanging around and the GIL is not an issue since the watcher mostly does sleeps and psycopg calls which release the GIL anyway.

File: click_odoo_contrib/update.py [hunk 6]
```diff
@@ -113,10 +217,20 @@ def OdooEnvironmentWithUpdate(database, ctx, **kwargs):
 @click.option(
     "--if-exists", is_flag=True, help="Don't report error if database doesn't exist"
 )
-def main(env, i18n_overwrite, update_all, if_exists):
+@click.option(
+    "--watcher-max-seconds",
+    default=0,
+    type=float,
+    help="Max DB lock seconds allowed before aborting the update process",
+)
```
[the remark this answers is not in the document]
Comment (Yajo) [reply click_odoo_contrib/update.py]: I prefer to have only 1 option if it is enough for the job.

File: click_odoo_contrib/update.py (hunk 4 above)
Comment (Yajo) [reply click_odoo_contrib/update.py]: The daemon process is supposed to be reaped automatically when the parent process exits, but the parent process is the current thread, so if we rely in automatic termination, it would leave the watcher working after the update has finished. It's not gonna do too much probably, but I prefer to terminate it in the exact moment when it's not needed anymore. The same applies for auto-closing threads.

About the GIL, I'm more worried the foreground Odoo update process locking the GIL, not actually the watcher.

This watcher is critical for CD/HA, so I prefer to go for the 100% secure way to handle it, even if it means having 4 LOC more. :smile:

**Comment by Yajo** [comment]
This is no longer WIP. :blush: 

It's working in production pretty well!

**Comment by blaggacao** [comment]
@Yajo It would be nice to include some explanation of this PR description in the readme. Additionally, it would be worthwhile to declare the covered and uncovered transitions and what might happen in each of those cases.

For example, if the schema changes are such that the database is unusable without the update (`res.users` schema change), then giving priority to users might just yield the undesirable.

It is a hard problem, insofar I think it's important to document the use cases and limitations extremely well.

For the before mentioned case, possible we would:
- Run update on new code while users still are served by old code.
- Hold a new code instance available.
- Switch over (L7) as soon as the update process terminates.
- Invalidate caches (?)

I can't get around the feeling that what we are trying to do here still requires a maintenance window in a good part of the cases, if the goal is to avoid 500 server errors.

If that's not the goal, then I think a valid question goes about the point of giving priority to the user.

Either way, giving priority to the user or to the update process will exclude some part of possible use cases to run cleanly.

If this is indeed to be a scoped solution in that sense, the limitations should be pointed out.

Note: cache invalidations is a different beast and probably only can be tackled at the operator level, but it's equally important for lim(seamless) updates.

File: click_odoo_contrib/update.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sbidoul) [reply click_odoo_contrib/update.py]: It's harder to explain but okay. Could you add a mention in the help text that 0 means the watcher is disabled?

File: click_odoo_contrib/update.py (hunk 4 above)
Comment (sbidoul) [reply click_odoo_contrib/update.py]: @Yajo I thought about this a bit more and did some tests with multiprocessing. My conclusion is that it's too easy to have the watcher process continuing after termination of the parent process. In my experiments, a SIGTERM to the parent is enough to leave the watcher running: that's sounds dangerous to me. 

A daemon thread will not have such issue.

In don't think the GIL would cause any issue in this case, as the GIL is relased with any postgress call.

File: click_odoo_contrib/update.py [hunk 7]
```diff
@@ -21,6 +25,87 @@
...
+                if locks:
+                    _logger.warning("%d locked queries found", len(locks))
+                    _logger.info("Query details: %r", locks)
+                for row in locks:
+                    if row["age"] > max_td:
+                        _logger.warning(
+                            "Long locked query detected; aborting update cursor..."
+                        )
+                        # Terminate the query to abort the parallel update
+                        watcher_cr.execute(stop_query, (row["pid"],))
```
Comment (sbidoul) [inline click_odoo_contrib/update.py]: Is it certain that killing row["pid"] will stop the update process and not the actual Odoo instance?

File: click_odoo_contrib/update.py [hunk 8]
```diff
@@ -113,10 +217,20 @@ def OdooEnvironmentWithUpdate(database, ctx, **kwargs):
...
+    help="Max DB lock seconds allowed before aborting the update process",
+)
+def main(env, i18n_overwrite, update_all, if_exists, watcher_max_seconds):
     """ Update an Odoo database (odoo -u), automatically detecting
     addons to update based on a hash of their file content, compared
     to the hashes stored in the database.
+
+    It allows updating in parallel while another Odoo instance is still
+    running against the same database, by using a watcher that aborts the
+    update in case a DB lock happens.
```
Comment (sbidoul) [inline click_odoo_contrib/update.py]: Can you update the relevant section in README with the new help text and the new options?

File: click_odoo_contrib/update.py (hunk 4 above)
Comment (Yajo) [reply click_odoo_contrib/update.py]: OK, I just changed it to thread. I think we should leave this open until I check there's no GIL problem.

**Comment by Yajo** [comment]
@blaggacao The scope of this improvement is to provide a tool as stated before. I don't want to document the best practices on how to use this in the wild here if possible, because there are many different deployment scenarios that can handle this differently. What it does is simple and explained, how do you use it is up to you.

Also keep in mind that Odoo is anti-HA by design. I'm just trying to get as close as possible. The main point here is to reduce downtime to almost nothing.

As always, I'm having a difficult time trying to understand your English! :sweat_smile:  Write simpler, direct questions instead, please, if anything else is left.

Thanks!

**Comment by blaggacao** [comment]
@Yajo Oh, I'm sorry for my English. Besides sharing my line of thought, the only actionable thing was the request of embedding the use cases in the the docs of this PR: for good code, I've read, that to document the _why_ is more important than the _what_, isn't it?

Also in order for others to be able to validate the _why_ (the "specification"), it needs to be documented somehow. As you might see, I'm still struggling to get a constructive critical hold (in a sense what Karl Popper stands for) on this. That's probably the root cause for my English is getting complicated. :wink:

**Comment by blaggacao** [comment]
If you feel I'm being too querulous here, please feel free tor ignore. Just know that my intention is Popperian. :stuck_out_tongue_winking_eye:

**Comment by Yajo** [comment]
I refurbished the script's exegenesis to keep it abreast of this new feature's natality motivation. I hope it appeases your popperian querulousity.

**Comment by blaggacao** [comment]
Exquisitely clear and tasty so!

In general: `sed '/lock/dead-lock/'` ? At least that's my understanding of it, now.

**Comment by Yajo** [comment]
I recently faced a failure where the lock watcher aborted a query but that allowed Odoo continue updating. I changed to use `pg_terminate_backend()` to close the cursor instead of `pg_cancel_backend()` which was just cancelling the query. I also added a system that will detect if the watcher aborted the main thread and will always return an error in such case.

**Comment by Yajo** [comment]
It seems [the failure is not related](https://travis-ci.org/acsone/click-odoo-contrib/jobs/557212775#L672) to this script.

This has been working fine in production for 5 months now. Last patches work pretty fine. Is there anything else left for it to be merged?

**Comment by sbidoul** [comment]
@Yajo yes, I'll need to fix the tests and then I'll merge. I removed the WIP label.

**Comment by Yajo** [comment]
:heart: Thanks!

**Comment by sbidoul** [comment]
@Yajo could you rebase? master is green.

**Comment by Yajo** [comment]
Done, thanks!

**Comment by Yajo** [comment]
So? It's green now :blush: :green_heart:

**Comment by sbidoul** [comment]
@Yajo I'm just back from holidays. What held me from clicking on merge before leaving was that I noticed the global variables which I did not see before (sorry about that). Do you think you could easily get rid of them, perhaps by implementing the watcher by subclassing `Thread` and overriding `run()`?

**Comment by Yajo** [comment]
I hope you enjoyed your holidays! :beach_umbrella: :sunglasses: 

The problem is that watcher thread needs to know if the main thread wants him to continue watching, and the main thread needs to know if the watcher thread has been aborted, and both react to conditions in the other one, so they need to share these 2 flags and [the easiest way I found was using a couple of global variables](https://stackoverflow.com/a/19790803/1468388).

How would a `Thread` subclass save us from that need? Also if that solution is more complicated, is it worth it? :thinking: Simpler is better IMHO...

**Comment by Yajo** [comment]
Thanks! Finally! :tada: Would you mind to release please?
